src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java

```diff
@@ -0,0 +1,364 @@
+package org.reactome.release.common.dataretrieval;	
+
+import java.io.ByteArrayOutputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.SocketException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.time.Duration;
+import java.time.Instant;
+
+import org.apache.commons.net.ftp.FTP;
+import org.apache.commons.net.ftp.FTPClient;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.conn.ConnectTimeoutException;
+import org.apache.http.conn.HttpHostConnectException;
+import org.apache.http.conn.UnsupportedSchemeException;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+
+public class FileRetriever implements DataRetriever {
+
+	protected URI uri;
+	protected String destination;
+	protected Duration maxAge;
+	protected Duration timeout = Duration.ofSeconds(30);
+	protected int numRetries = 1;
+	protected String retrieverName;
+	protected Logger logger;
+	protected boolean passiveFTP = false;
+	
+	
+	public FileRetriever()
+	{
+		this(null);
+	}
+	
+	/**
+	 * This constructor takes a string that will be used to name the logging output file.
+	 * 
+	 * @param retrieverName - the name of this File Retriever - the log file produced by this class will be named "${retrieverName}.log"
+	 */
+	public FileRetriever(String retrieverName)
+	{
+		//logger.trace("Setting retrieverName to {}", retrieverName);
+		this.setRetrieverName(retrieverName);
+		this.logger = this.createLogger(this.retrieverName, "RollingRandomAccessFile", this.getClass().getName(), true, Level.DEBUG, this.logger, "Data Retriever");
+	}
+	
+	@Override
+	public void fetchData() throws Exception 
+	{
+		if (this.uri == null)
+		{
+			throw new RuntimeException("You must provide a URI from which the file will be downloaded!");
```

Comment:
If you want, you could make this an IllegalStateException, so it is more specific

Reply:
For the next version of release-common-lib after this one, I was planning to put some time in for custom exception classes. The code is now an API that is getting more and more widely used so I think it's time to make that change. This would probably get fixed as a part of that.

Reply:
Okay, as long as it's on the radar.  Would it be before or after this PR is merged?  My concern is the comments getting lost showing the places for reworking if the changes are after this PR is merged

Reply:
After this PR is merged. _This_ PR is too big, already.

Reply:
Okay, you could just refer back to this PR for the comments to see what was suggested if needed.